src/Types.jl

```diff
@@ -892,52 +893,251 @@ function ensure_resolved(env::EnvCache,
+##############
+# Registries #
+##############
+
+mutable struct RegistrySpec
+    name::Union{String,Nothing}
+    uuid::Union{UUID,Nothing}
+    url::Union{String,Nothing}
+    # the path field can be a local source when adding a registry
+    # otherwise it is the path where the registry is installed
+    path::Union{String,Nothing}
+    RegistrySpec(name::String) = RegistrySpec(name = name)
+    RegistrySpec(;name=nothing, uuid=nothing, url=nothing, path=nothing) =
+        new(name, isa(uuid, String) ? UUID(uuid) : uuid, url, path)
+end
+
+const DEFAULT_REGISTRIES =
+    RegistrySpec[RegistrySpec(name = "General",
+                              uuid = UUID("23338594-aafe-5451-b93e-139f81909106"),
+                              url = "https://github.com/JuliaRegistries/General.git")]
 
-# Return paths of all registries in a depot
-function registries(depot::String)::Vector{String}
+# Return `RegistrySpec`s of each registry in a depot
+function collect_registries(depot::String)
     d = joinpath(depot, "registries")
-    ispath(d) || return String[]
-    regs = filter!(readdir(d)) do r
-        isfile(joinpath(d, r, "Registry.toml"))
+    regs = RegistrySpec[]
+    ispath(d) || return regs
+    for name in readdir(d)
+        for slug in readdir(joinpath(d, name))
+            file = joinpath(d, name, slug, "Registry.toml")
+            if isfile(file)
+                registry = read_registry(file)
+                # verify_registry(registry)
+                spec = RegistrySpec(name = registry["name"],
+                                    uuid = UUID(registry["uuid"]),
+                                    url = haskey(registry, "repo") ? registry["repo"] : nothing,
+                                    path = dirname(file))
+                push!(regs, spec)
+            end
+        end
     end
-    String[joinpath(depot, "registries", r) for r in regs]
+    return regs
 end
 
-# Return paths of all registries in all depots
-function registries(; clone_default=true)::Vector{String}
-    isempty(depots()) && return String[]
+# Return `RegistrySpec`s of all registries in all depots
+function collect_registries(; clone_default=true)
+    isempty(depots()) && return RegistrySpec[]
     user_regs = abspath(depots1(), "registries")
     if clone_default
         if !ispath(user_regs)
             mkpath(user_regs)
+            printpkgstyle(stdout, :Cloning, "default registries into $user_regs")
+            clone_or_cp_registries(DEFAULT_REGISTRIES, dirname(user_regs))
+        end
+    end
+    return RegistrySpec[r for d in depots() for r in collect_registries(d)]
+end
+
+function populate_known_registries_with_urls!(registries::Vector{RegistrySpec})
+    known_registries = DEFAULT_REGISTRIES # TODO: Some way to add stuff here?
+    for reg in registries, known in known_registries
+        if reg.uuid !== nothing
+            if reg.uuid === known.uuid
```

Comment:
`===` -> `==`?

Reply:
Also, I think it would be better to use `findfirst` so that you don't have to iterate over `known_registries` in the case that `reg.uuid === nothing`.

Reply:
```
julia> a = UUID("bdb5d00a-7a24-44d2-be8d-ca1835c73734");

julia> b = UUID("bdb5d00a-7a24-44d2-be8d-ca1835c73734");

julia> a === b
true
```

> Also, I think it would be better to use findfirst

You mean `findfirst(x -> reg.uuid == x.uuid, known_registries)` ?

Reply:
Ah, I didn't know `UUID` behaved like that, thanks.

And yes, that's what I meant by `findfirst`. It is a detail, so I think it's fine as is too.